libvirt: add e2e demo code and documentation (#345)
confidential-containers/cloud-api-adaptor: libvirt/README.md

with the kcli tool, as:

```
./kcli_cluster.sh create
```

With `kcli_cluster.sh` you can configure the libvirt network and storage pools that the cluster VMs will be created, among
other parameters. Run `./kcli_cluster.sh -h` to see the help for further information.

If everything goes well you will be able to see the cluster running after setting your Kubernetes config with:

`export KUBECONFIG=$HOME/.kcli/clusters/peer-pods/auth/kubeconfig`

For example, shown below:

```
$ kcli list kube
+-----------+---------+-----------+---------------------------------------+
|  Cluster  |   Type  |    Plan   |                  Vms                  |
+-----------+---------+-----------+---------------------------------------+
| peer-pods | generic | peer-pods | peer-pods-master-0,peer-pods-worker-0 |
+-----------+---------+-----------+---------------------------------------+
$ kubectl get nodes
NAME                 STATUS   ROLES                  AGE     VERSION
peer-pods-master-0   Ready    control-plane,master   6m8s    v1.25.3
peer-pods-worker-0   Ready    worker                 2m47s   v1.25.3
```

> ## Prepare the Pod VM volume

Reviewer: Can we move this section up to the top please. At the moment there is a far bit of duplication between this and the **Creation of base storage volume** section and if you are like me then we will have already spent far too long finding and installing the correct version of the pre-reqs and building the podvm image before you see the option to do it in the container.

Author: Talking about the podvm build in container, I opened an issue https://github.com/confidential-containers/cloud-api-adaptor/issues/550, it should be documented. I don't know if @bpradipt has already written about it somewhere, otherwise I can assign this task to myself.

I general i think libvirt's README should be re-write. So if you don't mind to merge this PR without major changes on it, then I can address the re-write in another PR.

Reviewer: Yes - I think that's fine given that you've created the follow on issue for it. Thanks

In order to build the Pod VM without installing the build tools, you can use the Dockerfiles hosted on [../podvm](../podvm) directory to run the entire process inside a container.

To build a Ubuntu VM image, run:

```
$ cd ../podvm
$ docker build --no-cache -t podvm_builder -f Dockerfile.podvm_builder .

> $ docker build --no-cache -t podvm_libvirt --build-arg BUILDER_IMG=podvm_builder:latest \

Reviewer: When running this I got an error in the libsecomp stage:
```
checking whether the C compiler works... no
configure: error: in `/tmp/build-libseccomp.Pv69G/libseccomp-2.5.4':
configure: error: C compiler cannot create executables
See `config.log' for more details
make: *** [Makefile.inc:100: /src/cloud-api-adaptor/podvm/staticlib/kata-libseccomp/lib/libseccomp.a] Error 77
```
I know it's not changed in the PR, but I thought it was worth noting.

Author: The ci-e2e_libvirt branch on my git tree misses a fix that I sent yesterday in another PR: https://github.com/confidential-containers/cloud-api-adaptor/pull/537

Please rebase your tree and then those commands should work out.

Author: Updated my tree.

Reviewer: I hit a timeout in the image creation again which was quite upsetting after waiting ~1 hour for it to run:
```
==> qemu.ubuntu: Waiting for SSH to become available...
==> qemu.ubuntu: Timeout waiting for SSH.
==> qemu.ubuntu: Deleting output directory...
Build 'qemu.ubuntu' errored after 5 minutes 57 seconds: Timeout waiting for SSH.

==> Wait completed after 5 minutes 57 seconds

==> Some builds didn't complete successfully and had errors:
--> qemu.ubuntu: Timeout waiting for SSH.

==> Builds finished but no artifacts were created.
make: *** [Makefile:69: podvm-f080e93-dirty-amd64.qcow2] Error 1
```

I tried updating `qcow2/ubuntu/qemu-ubuntu.pkr.hcl` to have `ssh_wait_timeout  = "600s"`, but I'm not sure if that's the correct setting as it didn't seem to help and ` 5 minutes 57 seconds` is longer than the default of 300s?
```
==> qemu.ubuntu: Timeout waiting for SSH.
==> qemu.ubuntu: Deleting output directory...
Build 'qemu.ubuntu' errored after 6 minutes 31 seconds: Timeout waiting for SSH.

==> Wait completed after 6 minutes 31 seconds

==> Some builds didn't complete successfully and had errors:
--> qemu.ubuntu: Timeout waiting for SSH.

==> Builds finished but no artifacts were created.
make: *** [Makefile:69: podvm-f080e93-dirty-amd64.qcow2] Error 1
```

Reviewer: It's worth noting that this passed for me in less than 4 mins when I ran it not in the container:
```
Build 'qemu.ubuntu' finished after 3 minutes 58 seconds.

==> Wait completed after 3 minutes 58 seconds
```

Reviewer 2: @wainersm, @stevenhorsman , i didn't hit this issue but one thing that may improve it is to pass `--device=/dev/kvm` (or even --privileged ) as it seems the VM starting inside the container is not accelerated with kvm, in any case worth mentioning it i think
Thanks to @whitebrandy who noticed this!

Reviewer: @snir911  hey Snir, sorry for being slow, but where do you mean to add `--device=/dev/kvm`. It looks like `docker build`, doesn't take that as an argument, just `docker run`?
```
podvm# docker build --no-cache --device=/dev/kvm -t podvm_builder -f Dockerfile.podvm_builder .
unknown flag: --device
See 'docker build --help'.
```

Reviewer 2: oh, podman does, so i took it for granted also docker :P

Reviewer: So good news and bad news on this I guess: I still can't get docker to work and the suggested flags don't exist, but I switched to podman (and added `--device=/dev/kvm`) and it all ran without an issue. 

I guess the question is what to do here:
- Change to documentation to require podman
- Add a note to say that if docker fails to try pdoman
- Try something else to get docker working

Reviewer: As another data point - I took a fresh VM and running podman without `--device=/dev/kvm` also fails to build the image, so on my set-up (4 vCPU, 8GB VM) it seems like that is mandatory to get success here.

Author: If you don't pass /dev/kvm then QEMU will run on emulated (TCG) mode, which is slower than hardware virtualization (KVM).  We cannot use KVM on cross-build nor on github jobs (e.g. https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/.github/workflows/podvm.yaml) because Github default runners don't support nested virt.

We can try to speed up the emulated mode. For example, use `q35` instead of `pc` machine type (https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/qcow2/ubuntu/variables.pkr.hcl#L56). And/or increase a little bit the requested memory (2GB) - https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/qcow2/ubuntu/variables.pkr.hcl#L31

On the podvm/README.md (see https://github.com/confidential-containers/cloud-api-adaptor/issues/550) document the environment constraints and required minimum memory/disk.

I don't like requiring podman. It seems more of our developers run docker.

Anyway, doesn't things shouldn't be addressed on this PR. @stevenhorsman @snir911 do you think it is good enough to get merged?

Reviewer: So for this particular issue of getting the podvm built in a container, I'm ok for the tidy up to be added to part of #550.

For the wider PR, I'm a bit more mixed. - I think it's a great start, but the fact that I've not been able to get the scripts working and produce a libvirt peer pod set-up does bother me. It suggests that there is either a problem in the code (unlikely), a problem in what I'm doing (likely, but that indicates there is possibly a documentation problem), or a issue in my set-up that is different (again not something that is probably not documented). My concern with merging it so that we can then build on it is if we'll end up in a situation that no-one outside of RedHat can get it working, which isn't a great place. If it is just me that can't get it working, then I'm fine with it though.

On the other hand if we can merge then, then start automating and improving some of the steps e.g. with ansible scripts, and get a libvirt CI/CD running in the next step after this in open-source then that would be a huge step and worth 'breaking some eggs' for.

Reviewer 2: @stevenhorsman I'm pretty sure it would work for you with more resources, for sure if it runs on baremetal and not in a VM (as the VM packer uses is nested and in addition it's not accelerated) . Maybe we can have a note mentioning this intensive operation that requires a lot of resources and shouldn't run in a VM or something.
BTW when i tried to run the build for RHEL in a VM it failed due to disk size, I'm currently trying to optimize it on this aspect (https://github.com/confidential-containers/cloud-api-adaptor/pull/525) IDK if we can improve it on the other aspects as well

Reviewer 2: It seems emulating x86_64 on m1 has lower performanc,. I think it worth trying 8vCPUs / 32GB if it's not a burden

Reviewer: FYI - the podvm build with docker worked fine in my 8 vCPU, 32GB RAM IBM Cloud VSI

Author: I tried within a VM with 4 vCPUs, 8GB Mem, and 30GB disk on my Linux Fedora 34 laptop. It worked out.

Of course the build still requires much disk (being addressed in https://github.com/confidential-containers/cloud-api-adaptor/pull/525) and packer within container takes a good amount of time because QEMU emulated. This last problem can be mitigated by passing /dev/kvm (i.e. allowing KVM acceleration on container). I will be documenting all this on the PR for https://github.com/confidential-containers/cloud-api-adaptor/issues/550

Reviewer 3: > @wainersm, @stevenhorsman , i didn't hit this issue but one thing that may improve it is to pass `--device=/dev/kvm` (or even --privileged ) as it seems the VM starting inside the container is not accelerated with kvm, in any case worth mentioning it i think Thanks to @whitebrandy who noticed this!

So, back to the original comment, the closest I know of is specifying "RUN --security=insecure" which is similar to running "docker run --privileged" but sorry, untested :)

	--build-arg CLOUD_PROVIDER=libvirt -f Dockerfile.podvm .
```

The qcow2 image file will be created inside the `podvm_libvirt` image and can be extracted as shown below:

```
$ podvm_container=$(docker create podvm_libvirt:latest /bin/bash)
$ file_name=$(docker export ${podvm_container} | tar -t | grep podvm)
$ docker cp ${podvm_container}:${file_name} podvm.qcow2
$ docker rm ${podvm_container}
```

Next you will need to create a volume on libvirt's system storage and upload the image content. That volume is used by
the cloud-api-adaptor program to instantiate a new Pod VM. Still on the `image` directory, you should run the following
commands:

```
$ export IMAGE=<full-path-to-qcow2>

$ virsh vol-create-as --pool default --name podvm-base.qcow2 --capacity 20G --allocation 2G --prealloc-metadata --format qcow2
$ virsh vol-upload --vol podvm-base.qcow2 $IMAGE --pool default --sparse
```

> You should see that the `podvm-base.qcow2` volume was proper created:

Reviewer 4: nit: proper->properly

Author: done

```
$ virsh -c qemu:///system vol-info --pool default podvm-base.qcow2
Name:           podvm-base.qcow2
Type:           file
Capacity:       6.00 GiB
Allocation:     631.52 MiB
```

## Install and configure Confidential Containers and cloud-api-adaptor in the cluster

The easiest way to install the cloud-api-adaptor along with Confidential Containers in the cluster is through the
Kubernetes operator available in the `install` directory of this repository.

Start by creating a public/private RSA key pair that will be used by the cloud-api-provider program, running on the
cluster workers, to connect with your local libvirtd instance without password authentication. Assume you are in the
`libvirt` directory, do:

```
$ cd ../install/overlays/libvirt
$ ssh-keygen -f ./id_rsa -N ""
$ cat id_rsa.pub >> ~/.ssh/authorized_keys
```

> **Note**: ensure that `~/.ssh/authorized_keys` has the right permissions (read/write for the user only) otherwise the

Reviewer: Is it worth suggest they do `chmod 600 ~/.ssh/authorized_keys` here to ensure this?

Author: Done.

authentication can silently fail.

You will need to figure out the IP address of your local host (e.g. 192.168.1.107). Then try to remote connect with
libvirt to check the keys setup is fine, for example:

```
$ virsh -c "qemu+ssh://$USER@192.168.1.107/system?keyfile=$(pwd)/id_rsa" nodeinfo
CPU model:           x86_64
CPU(s):              12
CPU frequency:       1084 MHz
CPU socket(s):       1
Core(s) per socket:  6
Thread(s) per core:  2
NUMA cell(s):        1
Memory size:         32600636 KiB
```

Now you should finally install the Kubernetes operator in the cluster with the help of the [`install_operator.sh`](./install_operator.sh) script. Ensure that you have your IP address exported in the environment, as shown below, then run the install script:

```
$ cd ../../../libvirt/
$ export LIBVIRT_IP="192.168.1.107"
$ export SSH_KEY_FILE="id_rsa"

> $ ./install_operator.sh

Reviewer: When I ran this I hit an error:
```
# ./install_operator.sh
kubectl apply -f install/yamls/deploy.yaml
namespace/confidential-containers-system created
serviceaccount/cc-operator-controller-manager created
role.rbac.authorization.k8s.io/cc-operator-leader-election-role created
clusterrole.rbac.authorization.k8s.io/cc-operator-manager-role created
clusterrole.rbac.authorization.k8s.io/cc-operator-metrics-reader created
clusterrole.rbac.authorization.k8s.io/cc-operator-proxy-role created
rolebinding.rbac.authorization.k8s.io/cc-operator-leader-election-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/cc-operator-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/cc-operator-proxy-rolebinding created
configmap/cc-operator-manager-config created
service/cc-operator-controller-manager-metrics-service created
deployment.apps/cc-operator-controller-manager created
customresourcedefinition.apiextensions.k8s.io/ccruntimes.confidentialcontainers.org created
kubectl apply -k install/overlays/libvirt
E0119 08:44:17.502494   56037 memcache.go:255] couldn't get resource list for confidentialcontainers.org/v1beta1: the server could not find the requested resource
configmap/peer-pods-cm-libvirt created
secret/auth-json-secret-libvirt created
secret/peer-pods-secret-libvirt created
secret/ssh-key-secret-libvirt created
daemonset.apps/cloud-api-adaptor-daemonset-libvirt created
error: resource mapping not found for name: "ccruntime-sample-libvirt" namespace: "confidential-containers-system" from "install/overlays/libvirt": no matches for kind "CcRuntime" in version "confidentialcontainers.org/v1beta1"
ensure CRDs are installed first
make: *** [Makefile:88: deploy] Error 1
```
and it finished very quickly, so I don't know if there should have been a wait in there or something? When I checked the pods they had been created:
```
# kubectl get pods -n confidential-containers-system
NAME                                             READY   STATUS    RESTARTS   AGE
cc-operator-controller-manager-fbb5dcf9d-w4jts   2/2     Running   0          59s
cloud-api-adaptor-daemonset-libvirt-j88xv        1/1     Running   0          56s
```
I re-ran it and it worked this time: `ccruntime.confidentialcontainers.org/ccruntime-sample-libvirt created`

were created on the cluster, as for example:

```
$ kubectl get runtimeclass
NAME        HANDLER     AGE
kata        kata        7m41s
kata-clh    kata-clh    7m41s
kata-qemu   kata-qemu   7m41s
```

## Create a sample peer-pods pod

At this point everything should be fine to get a sample Pod created. Let's first list the running VMs so that we can later check
the Pod VM will be really running. Notice below that we got only the cluster node VMs up:

```
$ virsh -c qemu:///system list
 Id   Name                 State
------------------------------------
 3    peer-pods-master-0   running
 4    peer-pods-worker-0   running
```

Create the *sample_busybox.yaml* file with the following content:

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: busybox
  name: busybox
spec:
  containers:
  - image: quay.io/prometheus/busybox
    name: busybox
    resources: {}
  dnsPolicy: ClusterFirst
  restartPolicy: Never
  runtimeClassName: kata
```

And create the Pod:

```
$ kubectl apply -f sample_busybox.yaml
pod/busybox created
$ kubectl wait --for=condition=Ready pod/busybox
pod/busybox condition met
```

> 

Reviewer: Creating the pod failed for me:
```
Events:
  Type     Reason                  Age   From               Message
  ----     ------                  ----  ----               -------
  Normal   Scheduled               62s   default-scheduler  Successfully assigned default/busybox to peer-pods-worker-0
  Warning  FailedMount             62s   kubelet            MountVolume.SetUp failed for volume "kube-api-access-67gfb" : failed to sync configmap cache: timed out waiting for the condition
  Normal   AddedInterface          57s   multus             Add eth0 [10.244.1.4/24] from cbr0
  Warning  FailedCreatePodSandBox  2s    kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: ttrpc: closed: unknown
```
It's probably user error, but I'm not sure how to resolve it.

Reviewer: I got a different error on this set-up:
```
Events:
  Type     Reason                  Age               From               Message
  ----     ------                  ----              ----               -------
  Normal   Scheduled               107s              default-scheduler  Successfully assigned default/busybox to peer-pods-worker-0
  Normal   AddedInterface          102s              multus             Add eth0 [10.244.1.12/24] from cbr0
  Warning  FailedCreatePodSandBox  61s               kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: ttrpc: closed: unknown
  Normal   AddedInterface          53s               multus             Add eth0 [10.244.1.13/24] from cbr0
  Normal   AddedInterface          41s               multus             Add eth0 [10.244.1.14/24] from cbr0
  Normal   AddedInterface          21s               multus             Add eth0 [10.244.1.15/24] from cbr0
  Normal   AddedInterface          12s               multus             Add eth0 [10.244.1.16/24] from cbr0
  Warning  FailedCreatePodSandBox  6s (x4 over 47s)  kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: failed to connect to remote hypervisor socket: dial unix /run/peerpod/hypervisor.sock: connect: connection refused: unknown
  Normal   AddedInterface          2s                multus             Add eth0 [10.244.1.17/24] from cbr0
```

One thing I'll note is that containerd doesn't seem to be installed, which I'd expect the pre-install section to do  when the operator is installed?
```
# containerd version

Command 'containerd' not found, but can be installed with:

apt install containerd
```

Reviewer: On my big boy 8vCPU 32GB RAM the libvirt peer pod worked! (almost). The `kubectl wait --for=condition=Ready pod/busybox` failed as it took just over 30seconds to be created, but I think if we can document that you might need a really big machine then I'm happy with this. Thanks for the patience all!

Check that the Pod VM is up and running. See on the following listing that *podvm-busybox-88a70031* was
created:

```
$ virsh -c qemu:///system list
 Id   Name                     State
----------------------------------------
 5    peer-pods-master-0       running
 6    peer-pods-worker-0       running
 7    podvm-busybox-88a70031   running
```

You should also check that the container is running fine. For example, compare the kernels are different as shown below:

```
$ uname -r
5.17.12-100.fc34.x86_64
$ kubectl exec pod/busybox -- uname -r
5.4.0-131-generic

> ```

Reviewer: I wonder if a clean-up/uninstall might be useful? That can be another PR though.

Author: Done.

Reviewer: Sorry, I wasn't particularly clear here. I was more thinking about the operator install, in case a user wanted to update (which I think would be an uninstall and then re-install, but if installing over the top works then this is fine)?

Author: ah, ok. I never installed a new version over the top, so no idea it works (@bpradipt ?). If so then it is a good use test case for the operator tests.

I will add some words about uninstall/install the operator.

Author: Done

Refer to the [cloud-api-adaptor deployment instructions](../install/README.md#deploy-cloud-api-adaptor)
* If your libvirt host is not configured to be passwordless make sure you set ssh-key-secret in the [kustomization file](../install/overlays/libvirt/kustomization.yaml)

# Creating an end-to-end environment for testing and development

In this section you will learn how to setup an environment in your local machine to run peer pods with
the libvirt cloud API adaptor. Bear in mind that many different tools can be used to setup the environment
and here we just make suggestions of tools that seems used by most of the peer pods developers.

## Requirements

You must have a Linux/KVM system with libvirt installed and the following tools:

- docker (or podman-docker)
- [kubectl](https://kubernetes.io/docs/reference/kubectl/)
- [kcli](https://kcli.readthedocs.io/en/latest/)

Assume that you have a 'default' network and storage pools created in libvirtd system instance (`qemu:///system`). However,
if you have a different pool name then the scripts should be able to handle it properly.

## Create the Kubernetes cluster

> Use the [`kcli_cluster.sh`](./kcli_cluster.sh) script to create a simple two VMs (one master and one worker) cluster

Reviewer: ```suggestion
Use the [`kcli_cluster.sh`](./kcli_cluster.sh) script to create a simple two VMs (one control plane and one worker) cluster
```

with the kcli tool, as:

```
./kcli_cluster.sh create
```

With `kcli_cluster.sh` you can configure the libvirt network and storage pools that the cluster VMs will be created, among
other parameters. Run `./kcli_cluster.sh -h` to see the help for further information.

If everything goes well you will be able to see the cluster running after setting your Kubernetes config with:

`export KUBECONFIG=$HOME/.kcli/clusters/peer-pods/auth/kubeconfig`

For example, shown below:

```
$ kcli list kube
+-----------+---------+-----------+---------------------------------------+
|  Cluster  |   Type  |    Plan   |                  Vms                  |
+-----------+---------+-----------+---------------------------------------+

> | peer-pods | generic | peer-pods | peer-pods-master-0,peer-pods-worker-0 |

Reviewer: ```suggestion
| peer-pods | generic | peer-pods | peer-pods-ctlplane-0,peer-pods-worker-0 |
```

+-----------+---------+-----------+---------------------------------------+
$ kubectl get nodes
NAME                 STATUS   ROLES                  AGE     VERSION

> peer-pods-master-0   Ready    control-plane,master   6m8s    v1.25.3

Reviewer: ```suggestion
peer-pods-ctlplane-0   Ready    control-plane,master   6m8s    v1.25.3
```

commands:

```
$ export IMAGE=<full-path-to-qcow2>

$ virsh vol-create-as --pool default --name podvm-base.qcow2 --capacity 20G --allocation 2G --prealloc-metadata --format qcow2
$ virsh vol-upload --vol podvm-base.qcow2 $IMAGE --pool default --sparse
```

You should see that the `podvm-base.qcow2` volume was properly created:

```
$ virsh -c qemu:///system vol-info --pool default podvm-base.qcow2
Name:           podvm-base.qcow2
Type:           file
Capacity:       6.00 GiB
Allocation:     631.52 MiB
```

## Install and configure Confidential Containers and cloud-api-adaptor in the cluster

The easiest way to install the cloud-api-adaptor along with Confidential Containers in the cluster is through the
Kubernetes operator available in the `install` directory of this repository.

Start by creating a public/private RSA key pair that will be used by the cloud-api-provider program, running on the
cluster workers, to connect with your local libvirtd instance without password authentication. Assume you are in the
`libvirt` directory, do:

```
$ cd ../install/overlays/libvirt
$ ssh-keygen -f ./id_rsa -N ""
$ cat id_rsa.pub >> ~/.ssh/authorized_keys
```

> **Note**: ensure that `~/.ssh/authorized_keys` has the right permissions (read/write for the user only) otherwise the

Reviewer 2: comment can be removed now?

Author: I didn't get. Why remove it? Mind to expand a little bit on your comment?

Reviewer 2: you do chmod no?

2022/11/09 18:18:00 [helper/hypervisor] hypervisor config {/run/peerpod/hypervisor.sock  k8s.gcr.io/pause:3.7 /run/peerpod/pods libvirt}
2022/11/09 18:18:00 [helper/hypervisor] cloud config {qemu+ssh://wmoschet@192.168.122.1/system?no_verify=1 default default /opt/data-dir}
2022/11/09 18:18:00 [helper/hypervisor] service config &{qemu+ssh://wmoschet@192.168.122.1/system?no_verify=1 default default /opt/data-dir}
```

You will also notice that Kubernetes [*runtimeClass*](https://kubernetes.io/docs/concepts/containers/runtime-class/) resources
were created on the cluster, as for example:

```
$ kubectl get runtimeclass
NAME        HANDLER     AGE
kata        kata        7m41s
kata-clh    kata-clh    7m41s
kata-qemu   kata-qemu   7m41s
```

## Create a sample peer-pods pod

At this point everything should be fine to get a sample Pod created. Let's first list the running VMs so that we can later check
the Pod VM will be really running. Notice below that we got only the cluster node VMs up:

```
$ virsh -c qemu:///system list
 Id   Name                 State
------------------------------------

>  3    peer-pods-master-0   running
>  4    peer-pods-worker-0   running

Reviewer: ```suggestion
 1    peer-pods-ctlplane-0   running
 2    peer-pods-worker-0     running
```

```

Create the *sample_busybox.yaml* file with the following content:

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: busybox
  name: busybox
spec:
  containers:
  - image: quay.io/prometheus/busybox
    name: busybox
    resources: {}
  dnsPolicy: ClusterFirst
  restartPolicy: Never
  runtimeClassName: kata
```

And create the Pod:

```
$ kubectl apply -f sample_busybox.yaml
pod/busybox created
$ kubectl wait --for=condition=Ready pod/busybox
pod/busybox condition met
```

Check that the Pod VM is up and running. See on the following listing that *podvm-busybox-88a70031* was
created:

```
$ virsh -c qemu:///system list
 Id   Name                     State
----------------------------------------

>  5    peer-pods-master-0       running
>  6    peer-pods-worker-0       running
>  7    podvm-busybox-88a70031   running

Reviewer: ```suggestion
 1    peer-pods-ctlplane-0     running
 2    peer-pods-worker-0       running
 3    podvm-busybox-41e459c1   running
```

```

You should also check that the container is running fine. For example, compare the kernels are different as shown below:

```
$ uname -r
5.17.12-100.fc34.x86_64
$ kubectl exec pod/busybox -- uname -r
5.4.0-131-generic
```

The peer-pods pod can be deleted as any regular pod. On the listing below the pod was removed and you can note that the
Pod VM no longer exists on Libvirt:

```
$ kubectl delete -f sample_busybox.yaml
pod "busybox" deleted
$ virsh -c qemu:///system list
 Id   Name                 State
------------------------------------

>  5    peer-pods-master-0   running
>  6    peer-pods-worker-0   running

Reviewer: ```suggestion
 1    peer-pods-ctlplane-0     running
 2    peer-pods-worker-0       running
```

```

# Creating an end-to-end environment for testing and development

In this section you will learn how to setup an environment in your local machine to run peer pods with

> the libvirt cloud API adaptor. Bear in mind that many different tools can be used to setup the environment

Reviewer 4: I think it's better to mention that this requires a Linux/KVM system with libvirt. Some devs do use Mac ;)

and here we just make suggestions of tools that seems used by most of the peer pods developers.

## Requirements

You must have installed in your local machine:

- docker (or podman-docker)
- [kcli](https://kcli.readthedocs.io/en/latest/)

> 

Author: That's a good idea @stevenhorsman ! I will come up with a separate PR; and on this one I will list kubectl as a requirement.

virsh -c qemu:///system vol-upload --vol podvm-base.qcow2 output/podvm-7da706d-dirty-amd64.qcow2 \
	--pool default --sparse
```

You should see that the `podvm-base.qcow2` volume was proper created:

```
$ virsh -c qemu:///system vol-info --pool default podvm-base.qcow2
Name:           podvm-base.qcow2
Type:           file
Capacity:       6.00 GiB
Allocation:     631.52 MiB
```

## Install and configure Confidential Containers and cloud-api-adaptor in the cluster

The easiest way to install the cloud-api-adaptor along with Confidential Containers in the cluster is through the
Kubernetes operator available in the `install` directory of this repository.

Start by creating a public/private RSA key pair that will be used by the cloud-api-provider program, running on the
cluster workers, to connect with your local libvirtd instance without password authentication. Assume you are in the
`libvirt` directory, do:

```
$ cd ../install/overlays/libvirt
$ ssh-keygen -f ./id_rsa -N ""
$ cat id_rsa.pub >> ~/.ssh/authorized_keys
```
**Note**: ensure that `~/.ssh/authorized_keys` has the right permissions (read/write for the user only) otherwise the
authentication can silently fail.

> You will need to figure out the IP address of your local host (e.g. 192.168.1.107). Then try to remote connect with

Reviewer 2: isn't it defaults to 192.168.122.1 ? maybe because you non-root?

Reviewer 4: default is usually 192.168.122.1. So may be the example can be changed to that.

Reviewer 4: But it's fine as-is as well.

Author: Done.

run the entire process inside a container.

Ensure that the Kata Containers repository is cloned side-by-side with the cloud-api-provider's, as both repositories will be
mounted as volumes in the container environment. Refer to [docs/DEVELOPMENT.md](/docs/DEVELOPMENT.md) for further
information about setup of the sources.

Then run:

```
$ cd image
$ make docker-build
```

The qcow2 image file will be created at the `output` directory as shown below:

```
$ ls output/
podvm-7da706d-dirty-amd64.qcow2
```

Next you will need to create a volume on libvirt's system storage and upload the image content. That volume is used by
the cloud-api-adaptor program to instantiate a new Pod VM. Still on the `image` directory, you should run `make push` just
like below:

```
$ make push
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5394  100  5394    0     0   6421      0 --:--:-- --:--:-- --:--:--  6421
virsh -c qemu:///system vol-create-as --pool default --name podvm-base.qcow2 \

> 	--capacity 107374182400 --allocation 2361393152 --prealloc-metadata \

Reviewer 2: nit: change it to match the` --capacity 20G --allocation 2G `change

Author: done
